### PR TITLE
[Ready] Cleaning up hypersleepers a little.

### LINKED ISSUE #987: Can't eject from hypersleep chamber.
## Reproduction: Get put in a hypersleep chamber, either by force or willingly, you cannot be removed or exit the hypersleep chamber, practically like putting someone out of the round by click dragging them.
dreamseeker_2019-03-05_18-30-45

### PR DESCRIPTION
**Pull request by Ghommie**
## About The Pull Request
Purges some indention, hypersleepers not using update_icon(), minor fixes. Closes #987 (Though it's still quite clunky to use verbs for this task). EDIT: Turns out I ended up reworking it extensively.

## Why It's Good For The Game
See above.

## Changelog
:cl:
refactor: Hypersleepers and their consoles got also reworked a little. The storage logs will also list the crewmember' job and the time they were hypersleeped. The list of retrievable items has been also incorporated within the console menu, rather than being a pop up input.
tweak: Expanded the list of items and trash the hypersleepers won't store away.
tweak: Hypersleep consoles can now access a list of cryo-storages rather a set one. Only affecting the Requisitions one right now.
add: Hypersleep consoles have been added to squad prep rooms for your squaddie scavenging need.
fix: Fixes hypersleeper occupants being unable to eject themselves.
/:cl:

### CODE REVIEW

File: code/game/machinery/cryopod.dm [hunk 1]
```diff
@@ -177,210 +177,211 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+			if("Master at Arms","Command Master at Arms")
+				dept_console = frozen_items["MP"]
+			if("Medical Officer","Medical Researcher","Chief Medical Officer")
+				dept_console = frozen_items["Med"]
+			if("Ship Engineer","Chief Ship Engineer")
+				dept_console = frozen_items["Eng"]
+
+	var/list/deleteempty = list(/obj/item/storage/backpack/marine/satchel)
+
+	var/list/deleteall = list(/obj/item/clothing/mask/cigarette, \
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: Could use being a define.

File: code/game/machinery/cryopod.dm [hunk 2]
```diff
@@ -177,210 +177,211 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+		var/mob/living/carbon/human/H = occupant
+		switch(H.job)
+			if("Master at Arms","Command Master at Arms")
+				dept_console = frozen_items["MP"]
+			if("Medical Officer","Medical Researcher","Chief Medical Officer")
+				dept_console = frozen_items["Med"]
+			if("Ship Engineer","Chief Ship Engineer")
+				dept_console = frozen_items["Eng"]
+
+	var/list/deleteempty = list(/obj/item/storage/backpack/marine/satchel)
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: Same here.

File: code/game/machinery/cryopod.dm [hunk 3]
```diff
@@ -177,210 +177,211 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+	/obj/item/clothing/glasses/sunglasses/sechud, \
+	/obj/item/device/radio/headset/almayer, \
+	/obj/item/card/id, \
+	/obj/item/clothing/under/marine, \
+	/obj/item/clothing/shoes/marine, \
+	/obj/item/clothing/head/cmcap)
+
+	var/list/strippeditems = list()
+
+	item_loop:
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: This god-awful thing needs to be made... Not so awful. Perhaps a separate proc?

File: code/game/machinery/cryopod.dm [hunk 4]
```diff
@@ -177,210 +177,211 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+		H.assigned_squad?.clean_marine_from_squad(H,TRUE) //Remove from squad recods, if any.
+
+	SSticker.mode.latejoin_tally-- //Cryoing someone out removes someone from the Marines, blocking further larva spawns until accounted for
+
+	//Handle job slot/tater cleanup.
+	if(occupant.mind?.assigned_role)
+		var/datum/job/J = SSjob.name_occupations[occupant.mind.assigned_role]
+		J.current_positions--
+
+	//Delete them from datacore.
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: This could also use to me atomized into a proc.

File: code/game/machinery/cryopod.dm [hunk 5]
```diff
@@ -177,210 +177,211 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+		for(var/obj/item/A in strippeditems)
+			for(var/DAA in deleteall)
+				if(istype(A, DAA))
+					qdel(A)
+					continue stripped_items
+
+			dept_console += A
+			A.loc = null
+
+	if(ishuman(occupant))
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: Same here

File: code/game/machinery/cryopod.dm [hunk 6]
```diff
@@ -177,210 +177,211 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+				if(istype(W, DA))
+					qdel(W)
+					continue item_loop
+
+
+
+			dept_console += W
+			W.loc = null
+
+	stripped_items:
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: This could also use a little bettering.

**Comment by Rohesie** [comment]
Could you make it send people back to lobby? Something like
```
var/mob/new_player/NP = new()
occupant.client.screen.Cut()
NP.key = occupant.key
NP.name = occupant.key
NP.client?.change_view(world.view)
```
Possibly tie that into a verb, or an option while ghosting.

File: code/game/machinery/cryopod.dm (hunk 1 above)
Comment (spookydonut) [reply code/game/machinery/cryopod.dm]: no, use a typecache

**Comment by Ghommie** [comment]
About the lobby thing, perhaps later. That alone would be exploitable, and it would be detrimental to have it ruled out solely through administration surveillance.

**Comment by Rohesie** [comment]
Fair enough.

File: code/game/machinery/cryopod.dm [hunk 7]
```diff
@@ -97,290 +96,248 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+		var/obj/item/I = input(usr, "Please choose which object to retrieve.", "Object recovery",null) as null|anything in stored_items
 
-		I.loc = get_turf(src)
-		frozen_items_for_type -= I
+		dispense_item(I)
 
 	else if(href_list["allitems"])
 
-		if(frozen_items_for_type.len == 0)
+		if(!length(stored_items.len))
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]:
```suggestion
		if(!length(stored_items))
```

File: code/game/machinery/cryopod.dm [hunk 8]
```diff
@@ -97,290 +96,248 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+/mob/proc/despawn(obj/machinery/cryopod/pod)
+	var/list/items = contents.Copy()
+
+	var/list/stored_items = list()
+
+	for(var/obj/item/W in items)
+		stored_items += W.place_in_storage()
+
+	var/list/dept_console = GLOB.cryoed_item_list["REQ"]
+	if(ishuman(src))
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: This should be in `/mob/living/carbon/human/despawn(obj/machinery/cryopod/pod)`.

File: code/game/machinery/cryopod.dm [hunk 9]
```diff
@@ -97,290 +96,248 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+					if(BRAVO_SQUAD)
+						dept_console = GLOB.cryoed_item_list["Bravo"]
+					if(CHARLIE_SQUAD)
+						dept_console = GLOB.cryoed_item_list["Charlie"]
+					if(DELTA_SQUAD)
+						dept_console = GLOB.cryoed_item_list["Delta"]
+
+	dept_console += stored_items
+
+	if(ishuman(src))
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: This should be in `/mob/living/carbon/human/despawn(obj/machinery/cryopod/pod)`.

File: code/game/machinery/cryopod.dm [hunk 10]
```diff
@@ -97,290 +96,248 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+					S.num_specialists--
+					if(H.specset && !available_specialist_sets.Find(H.specset))
+						available_specialist_sets += H.specset //we make the set this specialist took if any available again
+
+				if("Squad Smartgunner")
+					S.num_smartgun--
+				if("Squad Leader")
+					S.num_leaders--
+			S.count--
+		H.assigned_squad?.clean_marine_from_squad(H,TRUE) //Remove from squad recods, if any.
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: You can put that inside the `if(assigned_squad)` check once mind is out of the equation.

File: code/game/machinery/cryopod.dm [hunk 11]
```diff
@@ -97,290 +96,248 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+
+	SSticker.mode.latejoin_tally-- //Cryoing someone out removes someone from the Marines, blocking further larva spawns until accounted for
+
+	//Handle job slot/tater cleanup.
+	if(mind?.assigned_role)
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: Change this to `if(job)` as well.

File: code/game/machinery/cryopod.dm [hunk 12]
```diff
@@ -97,290 +96,248 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+		if((T.fields["name"] == real_name))
+			data_core.security -= T
+			qdel(T)
+	for(var/datum/data/record/G in data_core.general)
+		if((G.fields["name"] == real_name))
+			data_core.general -= G
+			qdel(G)
+
+
+	ghostize(0) //We want to make sure they are not kicked to lobby.
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]:
```suggestion
	ghostize(FALSE) //We want to make sure they are not kicked to lobby.
```

File: code/game/machinery/cryopod.dm [hunk 13]
```diff
@@ -97,290 +96,248 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+		return
+
+	if(is_type_in_typecache(src, GLOB.do_not_preserve))
+		qdel(src)
+		return
+
+	. += list(src)
+	//special items that store stuff in a nonstandard way, we properly remove those items
+
+	if(istype(src, /obj/item/clothing/suit/storage))
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: This `if` chain should go into proc redefinitions.

File: code/game/machinery/cryopod.dm [hunk 14]
```diff
@@ -81,14 +77,17 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
 	//	return
 
 	var/mob/user = usr
-	var/list/frozen_items_for_type = frozen_items[cryotype]
+	var/list/stored_items = GLOB.cryoed_item_list[cryotype]
+	for(var/obj/item/I in stored_items)
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: Is the typecheck here necessary?

File: code/game/machinery/cryopod.dm [hunk 15]
```diff
@@ -97,290 +96,259 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
+			data_core.security -= T
+			qdel(T)
+	for(var/datum/data/record/G in data_core.general)
+		if((G.fields["name"] == real_name))
+			data_core.general -= G
+			qdel(G)
+
+
+	ghostize(FALSE) //We want to make sure they are not kicked to lobby.
+	//TODO: Check objectives/mode, update new targets if this mob is the target, spawn new antags?
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: Pretty obsolete comment considering we have no syndicate code anymore

**Comment by Alterist** [comment]
seems like you didn't mapmerge properly

File: code/game/machinery/cryopod.dm [hunk 16]
```diff
@@ -466,56 +450,65 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
 	occupant = null
 	stop_processing()
-
-	if(orient_right)
-		icon_state = "body_scanner_0-r"
-	else
-		icon_state = "body_scanner_0"
+	update_icon()
+
+/obj/machinery/cryopod/admin //Invisible admin magic.
```
Comment (DominikPanic) [inline code/game/machinery/cryopod.dm]: This is 100x better, but the next step up would me to make it a mob proc so there's no need to spawn a machine.

**Comment by Ghommie** [comment]
> 
> 
> Also, what's the purpose of all of these map changes?

Implementing specific squad cryo consoles in the prep rooms.
Also carried on your latest request.

File: code/game/machinery/cryopod.dm [hunk 17]
```diff
@@ -77,310 +123,268 @@ var/global/list/frozen_items = list("Alpha"=list(),"Bravo"=list(),"Charlie"=list
...
-		icon_state = "body_scanner_0"
-	..()
+		radio = new(src)
+	radio.autosay("[real_name] has entered long-term hypersleep storage. Belongings moved to hypersleep inventory.", "Hypersleep Storage System")
+
+	qdel(src)
+
+/mob/living/carbon/human/despawn(obj/machinery/cryopod/pod, dept_console = CRYO_REQ)
+	if(assigned_squad)
+		var/datum/squad/S = assigned_squad
```
Comment (Rohesie) [inline code/game/machinery/cryopod.dm]: The `assigned_squad` mob var is already type cast, unlike the mind one. No need to do it again.

**Comment by Rohesie** [comment]
Is this locally tested?

**Comment by Ghommie** [comment]
> 
> 
> Is this locally tested?

Yes.

**Comment by Ghommie** [comment]
> 
> 
> Cryopods are missing a `start_processing` in their Initialize. This PR breaks the one thing it's supposed to be reworking.

Starting with the premise that cryopods should only start/stop processing when the occupant climbs in/out, I tested it again, after doing some minor adjustements, and were unable to reproduce this one bug and seems rather stable to me.

File: _maps/map_files/Theseus/TGS_Theseus.dmm [hunk 18]
```diff
@@ -15786,6 +15762,15 @@
...
 /area/almayer/engineering/upper_engineering)
+"aGT" = (
+/obj/effect/attach_point/electronics/dropship1{
+	dir = 1
+	},
+/turf/open/floor/plating,
+/turf/closed/shuttle/dropship1/transparent{
+	icon_state = "90"
+	},
+/area/shuttle/drop1/sulaco)
```
Comment (Rohesie) [inline _maps/map_files/Theseus/TGS_Theseus.dmm]: Are these changes intentional?

Comment (Ghommie) [reply _maps/map_files/Theseus/TGS_Theseus.dmm]: Seems like so, I mapped in one cryopod console for each squad in their prep rooms.

**Comment by Rohesie** [comment]
Tested locally?

**Comment by Ghommie** [comment]
> 
> 
> Tested locally?

Yes, did a roundabout on it, it seems fine.

**Comment by Rohesie** [comment]
Please give it a quick test after solving the conflicts.

**Comment by Ghommie** [comment]
> 
> 
> Please give it a quick test after solving the conflicts.

Done, already done. And the merge conflict was simply regarding the /device item subtype removal.

**Comment by Rohesie** [comment]
The do_after doesn't finish when you grab someone and try to put them in an hypersleep chamber.
This is what we call "lack of local testing".

**Comment by Ghommie** [comment]
Fixed. That turned out to be a typo I made during some after-test line changes. It's ok to re-TM it I assume.

**Pull request merged**